pkg/metrics/collector/kernel_metrics.go

```diff
@@ -0,0 +1,145 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package collector
+
+import (
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	NetstatLabel = "TcpExt:" // Label we use to parse netstat tcp metrics
+	SnmpLabel    = "Tcp:"    // Label we use to parse snmp tcp metrics
+)
+
+// Keys for metrics after parsing
+const (
+	// Got from Snmp counters
+	TCPSegsIn      = "InSegs"      // Total tcp segments in
+	TCPSegsOut     = "OutSegs"     // Total tcp segments out
+	TCPSegsRetrans = "RetransSegs" // Total tcp segments retransmitted
+	// Got from netstat
+	TCPTimeoutRehash   = "TcpTimeoutRehash"      // Total tcp timeout rehash
+	TCPDuplicateRehash = "TcpDupliateDataRehash" // Total tcp duplicates rehashed
+)
+
+var (
+	tcpTimeoutDesc = prometheus.NewDesc(
+		"tcp_timeout_rehash_count",
+		"Tcp timeout rehash count",
+		nil, nil,
+	)
+	tcpDuplicateDesc = prometheus.NewDesc(
+		"tcp_duplicate_rehash_count",
+		"Tcp duplicate rehash counts",
+		nil, nil,
+	)
+	segmentsReceivedDesc = prometheus.NewDesc(
+		"tcp_segments_received_count",
+		"TCP segments received on the node",
+		nil, nil,
+	)
+	segmentsSentDesc = prometheus.NewDesc(
+		"tcp_segments_sent_count",
+		"TCP segments sent from node",
+		nil, nil,
+	)
+	segmentsRetransmittedDesc = prometheus.NewDesc(
+		"tcp_segments_retransmitted_count",
+		"TCP segments retransmitted on the node",
+		nil, nil,
+	)
+)
+
+type kernelStatCollector struct {
+}
+
+func init() {
+	registerCollector("kernel_metrics", NewKernelStatCollector)
+}
+
+func NewKernelStatCollector() (Collector, error) {
+	return &kernelStatCollector{}, nil
+}
+
+// Parses a string which contains two lines, each starting with label
+// returng a map where the keys are the first lines elements, while the
+// second line contains the values, all casted to uint64, e.g
+//
+// label key1 key2 key3
+// label 1 2 3
+//
+// Outputs: {"key1":1 "key2":2 "key3":3}
+func parseKeyValueLines(output, label string) map[string]uint64 {
+	stats := make(map[string]uint64)
+	splitFile := strings.Split(output, "\n")
+	for i, stat := range splitFile {
+
+		if !strings.Contains(stat, label) {
+			continue
+		}
+		keyList := strings.Fields(stat)
+		// TODO: Look for next instance of label rather than assuming its the next
+		valList := strings.Fields(splitFile[i+1])
+		for i, key := range keyList[1:] {
+			val, err := strconv.Atoi(valList[i+1])
+			if err != nil {
+				continue
+			}
+			stats[key] = uint64(val)
+		}
+		return stats
+	}
+	return stats
+}
+
+func (c *kernelStatCollector) Update(ch chan<- prometheus.Metric) error {
+	// Get snmp values.
+	data, err := ioutil.ReadFile("/proc/net/snmp")
+	if err == nil {
```

Comment:
Would be great to keep the normal code path unindented: https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

I think it is fine to just return the error if `ReadFile` failed - the chance that it fails on `/proc/net/snmp` but suceeds on `/proc/net/netstat` should be low anyway.

Reply:
SG!